ocaml/xapi/xapi_database_backup.ml

```diff
@@ -0,0 +1,184 @@
+  {tblname=table; count=0; del=(Db_cache_types.Table.get_deleted_len t)}
+
+let object_count db =
+  let tables = get_table_list db in
+  List.map (counter db) tables
+
+let count_eq_check c1 c2 =
+  c1.tblname = c2.tblname && c1.count = c2.count && c1.del = c2.del
+
+let apply_changes (delta:delta) __context =
+  if (Db_cache_types.Manifest.generation (Db_cache_types.Database.manifest !Xapi_slave_db.slave_db)) < delta.fresh_token then
+    begin
+      Xapi_slave_db.slave_db := Database.set_generation delta.fresh_token !Xapi_slave_db.slave_db;
+      apply_updates __context delta.tables;
+      apply_deletes __context delta.deletes;
+    end;
+  (* Need to check that we haven't missed anything *)
+  if not (List.for_all (fun x -> x)
+      (List.map2 count_eq_check (object_count !Xapi_slave_db.slave_db) (object_count (Db_ref.get_database (Context.database_of __context))))) then
+    Mutex.execute Xapi_slave_db.slave_db_mutex (fun () -> Xapi_slave_db.slave_db := Db_cache_types.Database.set_generation 0L !Xapi_slave_db.slave_db)
+
+let info_extract (tlist:table list) =
+  let tbls = List.map (fun (i:table) -> i.tblname) tlist in
+  let rows = List.concat (List.map (fun tbl -> List.map (fun row -> row.objref) tbl.value) tlist) in
+  List.concat (List.map ((fun objlist tblname -> List.combine [tblname] objlist) rows) tbls)
+
+let handle_notifications __context (d:delta) =
+  List.iter ((fun __context (d:del_tables) -> send_notification __context "del" d.table d.key) __context) d.deletes;
+  let tbls, objs = List.split (info_extract d.tables) in
+  List.iter2 (send_notification __context "mod") tbls objs
+
+let write_db_to_disk =
+  debug "Writing slave db to slave disk";
+  Db_cache_impl.sync (Db_conn_store.read_db_connections ()) !Xapi_slave_db.slave_db
+
+let endless_loop ~__context () =
+  let uri = Constants.database_backup_uri in
+  let delay = Mtime.Span.of_uint64_ns 300_000_000_000L in
+  while (true) do
+    let now = ref (Mtime_clock.now ()) in
+    let addr = Pool_role.get_master_address() in
+    let psec = !Xapi_globs.pool_secret in
+    let token_str = Int64.to_string (Manifest.generation (Database.manifest !Xapi_slave_db.slave_db)) in
+    let req = Xapi_http.http_request ~cookie:["pool_secret", psec] ~query:[("token", token_str)] Http.Get uri in
+    let transport = SSL(SSL.make (), addr, !Xapi_globs.https_port) in
+    with_transport transport
+      (with_http req
+         (fun (response, fd) ->
+            let res = match response.Http.Response.content_length with
+              | None -> raise Content_length_required
+              | Some l -> Xapi_stdext_unix.Unixext.really_read_string fd (Int64.to_int l)
+            in
+            let delta_change = delta_of_rpc (Jsonrpc.of_string res) in
+            apply_changes delta_change __context;
+            handle_notifications __context delta_change;
+            if (Mtime.span (Mtime_clock.now ()) !now) > delay then
```

Comment:
I think you can use the `counter` function in Mtime to do this for you, you initialize it, and call `count` when you want to know how much has elapsed.